### ISSUE TITLE
Test failures on 32-bit systems

### ISSUE THREAD

**Issue by QuLogic** [dask/dask#2022]
I'm running [a build on multiple arches](https://copr.fedorainfracloud.org/coprs/qulogic/dask/build/519213/) and have run into consistent problems on 32-bit systems. Below are the tests that fail:
```
________________________________ test_no_chunks ________________________________
    def test_no_chunks():
        X = np.arange(11)
        dsk = {('x', 0): np.arange(5), ('x', 1): np.arange(5, 11)}
        x = Array(dsk, 'x', ((np.nan, np.nan,),), np.arange(1).dtype)
        assert_eq(x + 1, X + 1)
        assert_eq(x.sum(), X.sum())
        assert_eq((x + 1).std(), (X + 1).std())
        assert_eq((x + x).std(), (X + X).std())
>       assert_eq((x + x).std(keepdims=True), (X + X).std(keepdims=True))
dask/array/tests/test_array_core.py:2551: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
a = array([ 10.02888391]), b = array([ 6.32455532]), kwargs = {}
adt = dtype('float64'), bdt = dtype('float64')
    def assert_eq(a, b, **kwargs):
        if isinstance(a, Array):
            adt = a.dtype
            _check_dsk(a.dask)
            a = a.compute(get=get_sync)
            _maybe_check_dtype(a, adt)
        else:
            adt = getattr(a, 'dtype', None)
        if isinstance(b, Array):
            bdt = b.dtype
            _check_dsk(b.dask)
            assert bdt is not None
            b = b.compute(get=get_sync)
            _maybe_check_dtype(b, bdt)
        else:
            bdt = getattr(b, 'dtype', None)
    
        if str(adt) != str(bdt):
            diff = difflib.ndiff(str(adt).splitlines(), str(bdt).splitlines())
            raise AssertionError('string repr are different' + os.linesep +
                                 os.linesep.join(diff))
    
        try:
            if _not_empty(a) and _not_empty(b):
                # Treat all empty arrays as equivalent
                assert a.shape == b.shape
>               assert allclose(a, b, **kwargs)
E               assert allclose(array([ 10.02888391]), array([ 6.32455532]), **{})
dask/array/utils.py:73: AssertionError


____________________________ test_reductions_1D[i4] ____________________________
dtype = 'i4'
    @pytest.mark.parametrize('dtype', ['f4', 'i4'])
    def test_reductions_1D(dtype):
        x = np.arange(5).astype(dtype)
        a = da.from_array(x, chunks=(2,))
    
        reduction_1d_test(da.sum, a, np.sum, x)
        reduction_1d_test(da.prod, a, np.prod, x)
        reduction_1d_test(da.mean, a, np.mean, x)
>       reduction_1d_test(da.var, a, np.var, x)
dask/array/tests/test_reductions.py:60: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
dask/array/tests/test_reductions.py:35: in reduction_1d_test
    assert_eq(da_func(darr, keepdims=True), np_func(narr, keepdims=True))
dask/array/tests/test_reductions.py:21: in assert_eq
    _assert_eq(a, b, equal_nan=True)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
a = array([ 4.79]), b = array([ 2.]), kwargs = {'equal_nan': True}
adt = dtype('float64'), bdt = dtype('float64')
    def assert_eq(a, b, **kwargs):
        if isinstance(a, Array):
            adt = a.dtype
            _check_dsk(a.dask)
            a = a.compute(get=get_sync)
            _maybe_check_dtype(a, adt)
        else:
            adt = getattr(a, 'dtype', None)
        if isinstance(b, Array):
            bdt = b.dtype
            _check_dsk(b.dask)
            assert bdt is not None
            b = b.compute(get=get_sync)
            _maybe_check_dtype(b, bdt)
        else:
            bdt = getattr(b, 'dtype', None)
    
        if str(adt) != str(bdt):
            diff = difflib.ndiff(str(adt).splitlines(), str(bdt).splitlines())
            raise AssertionError('string repr are different' + os.linesep +
                                 os.linesep.join(diff))
    
        try:
            if _not_empty(a) and _not_empty(b):
                # Treat all empty arrays as equivalent
                assert a.shape == b.shape
>               assert allclose(a, b, **kwargs)
E               assert allclose(array([ 4.79]), array([ 2.]), **{'equal_nan': True})
dask/array/utils.py:73: AssertionError


____________________________ test_reductions_2D[i4] ____________________________
dtype = 'i4'
    @pytest.mark.parametrize('dtype', ['f4', 'i4'])
    def test_reductions_2D(dtype):
        x = np.arange(1, 122).reshape((11, 11)).astype(dtype)
        a = da.from_array(x, chunks=(4, 4))
    
        b = a.sum(keepdims=True)
        assert b._keys() == [[(b.name, 0, 0)]]
    
        reduction_2d_test(da.sum, a, np.sum, x)
        reduction_2d_test(da.prod, a, np.prod, x)
        reduction_2d_test(da.mean, a, np.mean, x)
>       reduction_2d_test(da.var, a, np.var, x, False)  # Difference in dtype algo
dask/array/tests/test_reductions.py:130: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
dask/array/tests/test_reductions.py:79: in reduction_2d_test
    assert_eq(da_func(darr, keepdims=True), np_func(narr, keepdims=True))
dask/array/tests/test_reductions.py:21: in assert_eq
    _assert_eq(a, b, equal_nan=True)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
a = array([[ 1722.79502766]]), b = array([[ 1220.]])
kwargs = {'equal_nan': True}, adt = dtype('float64'), bdt = dtype('float64')
    def assert_eq(a, b, **kwargs):
        if isinstance(a, Array):
            adt = a.dtype
            _check_dsk(a.dask)
            a = a.compute(get=get_sync)
            _maybe_check_dtype(a, adt)
        else:
            adt = getattr(a, 'dtype', None)
        if isinstance(b, Array):
            bdt = b.dtype
            _check_dsk(b.dask)
            assert bdt is not None
            b = b.compute(get=get_sync)
            _maybe_check_dtype(b, bdt)
        else:
            bdt = getattr(b, 'dtype', None)
    
        if str(adt) != str(bdt):
            diff = difflib.ndiff(str(adt).splitlines(), str(bdt).splitlines())
            raise AssertionError('string repr are different' + os.linesep +
                                 os.linesep.join(diff))
    
        try:
            if _not_empty(a) and _not_empty(b):
                # Treat all empty arrays as equivalent
                assert a.shape == b.shape
>               assert allclose(a, b, **kwargs)
E               assert allclose(array([[ 1722.79502766]]), array([[ 1220.]]), **{'equal_nan': True})
dask/array/utils.py:73: AssertionError


______________________ test_arg_reductions[dfunc0-func0] _______________________
dfunc = <function argmin at 0xf4a5472c>, func = <function argmin at 0xf6496b1c>
    @pytest.mark.parametrize(['dfunc', 'func'],
                             [(da.argmin, np.argmin), (da.argmax, np.argmax),
                              (da.nanargmin, np.nanargmin),
                              (da.nanargmax, np.nanargmax)])
    def test_arg_reductions(dfunc, func):
        x = np.random.random((10, 10, 10))
        a = da.from_array(x, chunks=(3, 4, 5))
    
>       assert_eq(dfunc(a), func(x))
dask/array/tests/test_reductions.py:154: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
dask/array/tests/test_reductions.py:21: in assert_eq
    _assert_eq(a, b, equal_nan=True)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
a = 265, b = 265, kwargs = {'equal_nan': True}, adt = dtype('int64')
bdt = dtype('int32'), diff = <generator object compare at 0xedcd920c>
    def assert_eq(a, b, **kwargs):
        if isinstance(a, Array):
            adt = a.dtype
            _check_dsk(a.dask)
            a = a.compute(get=get_sync)
            _maybe_check_dtype(a, adt)
        else:
            adt = getattr(a, 'dtype', None)
        if isinstance(b, Array):
            bdt = b.dtype
            _check_dsk(b.dask)
            assert bdt is not None
            b = b.compute(get=get_sync)
            _maybe_check_dtype(b, bdt)
        else:
            bdt = getattr(b, 'dtype', None)
    
        if str(adt) != str(bdt):
            diff = difflib.ndiff(str(adt).splitlines(), str(bdt).splitlines())
            raise AssertionError('string repr are different' + os.linesep +
>                                os.linesep.join(diff))
E           AssertionError: string repr are different
E           - int64
E           + int32
dask/array/utils.py:67: AssertionError


______________________ test_arg_reductions[dfunc1-func1] _______________________
dfunc = <function argmax at 0xf4a54764>, func = <function argmax at 0xf6496ae4>
    @pytest.mark.parametrize(['dfunc', 'func'],
                             [(da.argmin, np.argmin), (da.argmax, np.argmax),
                              (da.nanargmin, np.nanargmin),
                              (da.nanargmax, np.nanargmax)])
    def test_arg_reductions(dfunc, func):
        x = np.random.random((10, 10, 10))
        a = da.from_array(x, chunks=(3, 4, 5))
    
>       assert_eq(dfunc(a), func(x))
dask/array/tests/test_reductions.py:154: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
dask/array/tests/test_reductions.py:21: in assert_eq
    _assert_eq(a, b, equal_nan=True)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
a = 305, b = 305, kwargs = {'equal_nan': True}, adt = dtype('int64')
bdt = dtype('int32'), diff = <generator object compare at 0xf0ffb98c>
    def assert_eq(a, b, **kwargs):
        if isinstance(a, Array):
            adt = a.dtype
            _check_dsk(a.dask)
            a = a.compute(get=get_sync)
            _maybe_check_dtype(a, adt)
        else:
            adt = getattr(a, 'dtype', None)
        if isinstance(b, Array):
            bdt = b.dtype
            _check_dsk(b.dask)
            assert bdt is not None
            b = b.compute(get=get_sync)
            _maybe_check_dtype(b, bdt)
        else:
            bdt = getattr(b, 'dtype', None)
    
        if str(adt) != str(bdt):
            diff = difflib.ndiff(str(adt).splitlines(), str(bdt).splitlines())
            raise AssertionError('string repr are different' + os.linesep +
>                                os.linesep.join(diff))
E           AssertionError: string repr are different
E           - int64
E           + int32
dask/array/utils.py:67: AssertionError


______________________ test_arg_reductions[dfunc2-func2] _______________________
dfunc = <function _nanargmin at 0xf4a5479c>
func = <function nanargmin at 0xf67293e4>
    @pytest.mark.parametrize(['dfunc', 'func'],
                             [(da.argmin, np.argmin), (da.argmax, np.argmax),
                              (da.nanargmin, np.nanargmin),
                              (da.nanargmax, np.nanargmax)])
    def test_arg_reductions(dfunc, func):
        x = np.random.random((10, 10, 10))
        a = da.from_array(x, chunks=(3, 4, 5))
    
>       assert_eq(dfunc(a), func(x))
dask/array/tests/test_reductions.py:154: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
dask/array/tests/test_reductions.py:21: in assert_eq
    _assert_eq(a, b, equal_nan=True)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
a = 677, b = 677, kwargs = {'equal_nan': True}, adt = dtype('int64')
bdt = dtype('int32'), diff = <generator object compare at 0xe5a2207c>
    def assert_eq(a, b, **kwargs):
        if isinstance(a, Array):
            adt = a.dtype
            _check_dsk(a.dask)
            a = a.compute(get=get_sync)
            _maybe_check_dtype(a, adt)
        else:
            adt = getattr(a, 'dtype', None)
        if isinstance(b, Array):
            bdt = b.dtype
            _check_dsk(b.dask)
            assert bdt is not None
            b = b.compute(get=get_sync)
            _maybe_check_dtype(b, bdt)
        else:
            bdt = getattr(b, 'dtype', None)
    
        if str(adt) != str(bdt):
            diff = difflib.ndiff(str(adt).splitlines(), str(bdt).splitlines())
            raise AssertionError('string repr are different' + os.linesep +
>                                os.linesep.join(diff))
E           AssertionError: string repr are different
E           - int64
E           + int32
dask/array/utils.py:67: AssertionError


______________________ test_arg_reductions[dfunc3-func3] _______________________
dfunc = <function _nanargmax at 0xf4a547d4>
func = <function nanargmax at 0xf672941c>
    @pytest.mark.parametrize(['dfunc', 'func'],
                             [(da.argmin, np.argmin), (da.argmax, np.argmax),
                              (da.nanargmin, np.nanargmin),
                              (da.nanargmax, np.nanargmax)])
    def test_arg_reductions(dfunc, func):
        x = np.random.random((10, 10, 10))
        a = da.from_array(x, chunks=(3, 4, 5))
    
>       assert_eq(dfunc(a), func(x))
dask/array/tests/test_reductions.py:154: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
dask/array/tests/test_reductions.py:21: in assert_eq
    _assert_eq(a, b, equal_nan=True)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
a = 909, b = 909, kwargs = {'equal_nan': True}, adt = dtype('int64')
bdt = dtype('int32'), diff = <generator object compare at 0xf103fd9c>
    def assert_eq(a, b, **kwargs):
        if isinstance(a, Array):
            adt = a.dtype
            _check_dsk(a.dask)
            a = a.compute(get=get_sync)
            _maybe_check_dtype(a, adt)
        else:
            adt = getattr(a, 'dtype', None)
        if isinstance(b, Array):
            bdt = b.dtype
            _check_dsk(b.dask)
            assert bdt is not None
            b = b.compute(get=get_sync)
            _maybe_check_dtype(b, bdt)
        else:
            bdt = getattr(b, 'dtype', None)
    
        if str(adt) != str(bdt):
            diff = difflib.ndiff(str(adt).splitlines(), str(bdt).splitlines())
            raise AssertionError('string repr are different' + os.linesep +
>                                os.linesep.join(diff))
E           AssertionError: string repr are different
E           - int64
E           + int32
dask/array/utils.py:67: AssertionError


_____________________ test_nanarg_reductions[dfunc0-func0] _____________________
dfunc = <function _nanargmin at 0xf4a5479c>
func = <function nanargmin at 0xf67293e4>
    @pytest.mark.parametrize(['dfunc', 'func'],
                             [(da.nanargmin, np.nanargmin),
                              (da.nanargmax, np.nanargmax)])
    def test_nanarg_reductions(dfunc, func):
        x = np.random.random((10, 10, 10))
        x[5] = np.nan
        a = da.from_array(x, chunks=(3, 4, 5))
>       assert_eq(dfunc(a), func(x))
dask/array/tests/test_reductions.py:181: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
dask/array/tests/test_reductions.py:21: in assert_eq
    _assert_eq(a, b, equal_nan=True)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
a = 451, b = 451, kwargs = {'equal_nan': True}, adt = dtype('int64')
bdt = dtype('int32'), diff = <generator object compare at 0xf0fc748c>
    def assert_eq(a, b, **kwargs):
        if isinstance(a, Array):
            adt = a.dtype
            _check_dsk(a.dask)
            a = a.compute(get=get_sync)
            _maybe_check_dtype(a, adt)
        else:
            adt = getattr(a, 'dtype', None)
        if isinstance(b, Array):
            bdt = b.dtype
            _check_dsk(b.dask)
            assert bdt is not None
            b = b.compute(get=get_sync)
            _maybe_check_dtype(b, bdt)
        else:
            bdt = getattr(b, 'dtype', None)
    
        if str(adt) != str(bdt):
            diff = difflib.ndiff(str(adt).splitlines(), str(bdt).splitlines())
            raise AssertionError('string repr are different' + os.linesep +
>                                os.linesep.join(diff))
E           AssertionError: string repr are different
E           - int64
E           + int32
dask/array/utils.py:67: AssertionError


_____________________ test_nanarg_reductions[dfunc1-func1] _____________________
dfunc = <function _nanargmax at 0xf4a547d4>
func = <function nanargmax at 0xf672941c>
    @pytest.mark.parametrize(['dfunc', 'func'],
                             [(da.nanargmin, np.nanargmin),
                              (da.nanargmax, np.nanargmax)])
    def test_nanarg_reductions(dfunc, func):
        x = np.random.random((10, 10, 10))
        x[5] = np.nan
        a = da.from_array(x, chunks=(3, 4, 5))
>       assert_eq(dfunc(a), func(x))
dask/array/tests/test_reductions.py:181: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
dask/array/tests/test_reductions.py:21: in assert_eq
    _assert_eq(a, b, equal_nan=True)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
a = 634, b = 634, kwargs = {'equal_nan': True}, adt = dtype('int64')
bdt = dtype('int32'), diff = <generator object compare at 0xe5a22bbc>
    def assert_eq(a, b, **kwargs):
        if isinstance(a, Array):
            adt = a.dtype
            _check_dsk(a.dask)
            a = a.compute(get=get_sync)
            _maybe_check_dtype(a, adt)
        else:
            adt = getattr(a, 'dtype', None)
        if isinstance(b, Array):
            bdt = b.dtype
            _check_dsk(b.dask)
            assert bdt is not None
            b = b.compute(get=get_sync)
            _maybe_check_dtype(b, bdt)
        else:
            bdt = getattr(b, 'dtype', None)
    
        if str(adt) != str(bdt):
            diff = difflib.ndiff(str(adt).splitlines(), str(bdt).splitlines())
            raise AssertionError('string repr are different' + os.linesep +
>                                os.linesep.join(diff))
E           AssertionError: string repr are different
E           - int64
E           + int32
dask/array/utils.py:67: AssertionError


___________________________ test_reductions_2D_nans ____________________________
    def test_reductions_2D_nans():
        # chunks are a mix of some/all/no NaNs
        x = np.full((4, 4), np.nan)
        x[:2, :2] = np.array([[1, 2], [3, 4]])
        x[2, 2] = 5
        x[3, 3] = 6
        a = da.from_array(x, chunks=(2, 2))
    
        reduction_2d_test(da.sum, a, np.sum, x, False, False)
        reduction_2d_test(da.prod, a, np.prod, x, False, False)
        reduction_2d_test(da.mean, a, np.mean, x, False, False)
        reduction_2d_test(da.var, a, np.var, x, False, False)
        reduction_2d_test(da.std, a, np.std, x, False, False)
        reduction_2d_test(da.min, a, np.min, x, False, False)
        reduction_2d_test(da.max, a, np.max, x, False, False)
        reduction_2d_test(da.any, a, np.any, x, False, False)
        reduction_2d_test(da.all, a, np.all, x, False, False)
    
        reduction_2d_test(da.nansum, a, np.nansum, x, False, False)
        reduction_2d_test(da.nanprod, a, nanprod, x, False, False)
        reduction_2d_test(da.nanmean, a, np.nanmean, x, False, False)
>       reduction_2d_test(da.nanvar, a, np.nanvar, x, False, False)
dask/array/tests/test_reductions.py:216: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
dask/array/tests/test_reductions.py:79: in reduction_2d_test
    assert_eq(da_func(darr, keepdims=True), np_func(narr, keepdims=True))
dask/array/tests/test_reductions.py:21: in assert_eq
    _assert_eq(a, b, equal_nan=True)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
a = array([[ 8.02777778]]), b = array([[ 2.91666667]])
kwargs = {'equal_nan': True}, adt = dtype('float64'), bdt = dtype('float64')
    def assert_eq(a, b, **kwargs):
        if isinstance(a, Array):
            adt = a.dtype
            _check_dsk(a.dask)
            a = a.compute(get=get_sync)
            _maybe_check_dtype(a, adt)
        else:
            adt = getattr(a, 'dtype', None)
        if isinstance(b, Array):
            bdt = b.dtype
            _check_dsk(b.dask)
            assert bdt is not None
            b = b.compute(get=get_sync)
            _maybe_check_dtype(b, bdt)
        else:
            bdt = getattr(b, 'dtype', None)
    
        if str(adt) != str(bdt):
            diff = difflib.ndiff(str(adt).splitlines(), str(bdt).splitlines())
            raise AssertionError('string repr are different' + os.linesep +
                                 os.linesep.join(diff))
    
        try:
            if _not_empty(a) and _not_empty(b):
                # Treat all empty arrays as equivalent
                assert a.shape == b.shape
>               assert allclose(a, b, **kwargs)
E               assert allclose(array([[ 8.02777778]]), array([[ 2.91666667]]), **{'equal_nan': True})
dask/array/utils.py:73: AssertionError
----------------------------- Captured stderr call -----------------------------
/builddir/build/BUILD/dask-0.14.0/dask/array/reductions.py:269: RuntimeWarning: invalid value encountered in true_divide
  u = total / n
/builddir/build/BUILD/dask-0.14.0/dask/array/numpy_compat.py:45: RuntimeWarning: invalid value encountered in divide
  x = np.divide(x1, x2, out)


______________________ test_reductions_with_negative_axes ______________________
    def test_reductions_with_negative_axes():
        x = np.random.random((4, 4, 4))
        a = da.from_array(x, chunks=2)
    
>       assert_eq(a.argmin(axis=-1), x.argmin(axis=-1))
dask/array/tests/test_reductions.py:262: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
dask/array/tests/test_reductions.py:21: in assert_eq
    _assert_eq(a, b, equal_nan=True)
dask/array/utils.py:52: in assert_eq
    _maybe_check_dtype(a, adt)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
a = array([[3, 1, 0, 0],
       [1, 1, 2, 2],
       [2, 0, 2, 2],
       [2, 0, 0, 2]])
dtype = dtype('int64')
    def _maybe_check_dtype(a, dtype=None):
        # Only check dtype matches for non-empty
        if _not_empty(a):
>           assert a.dtype == dtype
E           assert dtype('int32') == dtype('int64')
E            +  where dtype('int32') = array([[3, 1, 0, 0],\n       [1, 1, 2, 2],\n       [2, 0, 2, 2],\n       [2, 0, 0, 2]]).dtype
dask/array/utils.py:33: AssertionError


___________________________________ test_nan ___________________________________
    def test_nan():
        x = np.array([[1, np.nan, 3, 4],
                      [5, 6, 7, np.nan],
                      [9, 10, 11, 12]])
        d = da.from_array(x, chunks=(2, 2))
    
        assert_eq(np.nansum(x), da.nansum(d))
        assert_eq(np.nansum(x, axis=0), da.nansum(d, axis=0))
        assert_eq(np.nanmean(x, axis=1), da.nanmean(d, axis=1))
        assert_eq(np.nanmin(x, axis=1), da.nanmin(d, axis=1))
        assert_eq(np.nanmax(x, axis=(0, 1)), da.nanmax(d, axis=(0, 1)))
        assert_eq(np.nanvar(x), da.nanvar(d))
        assert_eq(np.nanstd(x, axis=0), da.nanstd(d, axis=0))
>       assert_eq(np.nanargmin(x, axis=0), da.nanargmin(d, axis=0))
dask/array/tests/test_reductions.py:282: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
dask/array/tests/test_reductions.py:21: in assert_eq
    _assert_eq(a, b, equal_nan=True)
dask/array/utils.py:60: in assert_eq
    _maybe_check_dtype(b, bdt)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
a = array([0, 1, 0, 0]), dtype = dtype('int64')
    def _maybe_check_dtype(a, dtype=None):
        # Only check dtype matches for non-empty
        if _not_empty(a):
>           assert a.dtype == dtype
E           assert dtype('int32') == dtype('int64')
E            +  where dtype('int32') = array([0, 1, 0, 0]).dtype
dask/array/utils.py:33: AssertionError


_____________________________ test_cov_corr_stable _____________________________
    @pytest.mark.slow
    def test_cov_corr_stable():
        df = pd.DataFrame(np.random.random((20000000, 2)) * 2 - 1, columns=['a', 'b'])
        ddf = dd.from_pandas(df, npartitions=50)
>       assert_eq(ddf.cov(split_every=8), df.cov())
dask/dataframe/tests/test_dataframe.py:1945: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
dask/dataframe/utils.py:509: in assert_eq
    tm.assert_frame_equal(a, b, **kwargs)
/usr/lib/python2.7/site-packages/pandas/util/testing.py:1313: in assert_frame_equal
    obj='DataFrame.iloc[:, {0}]'.format(i))
/usr/lib/python2.7/site-packages/pandas/util/testing.py:1181: in assert_series_equal
    obj='{0}'.format(obj))
pandas/src/testing.pyx:59: in pandas._testing.assert_almost_equal (pandas/src/testing.c:3783)
    ???
pandas/src/testing.pyx:163: in pandas._testing.assert_almost_equal (pandas/src/testing.c:2741)
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
>   ???
E   AssertionError: expected 0.00001 but got 0.00001, with decimal 5
pandas/src/testing.pyx:209: AssertionError
```
Presumably some of these tests are using the default `float` with some internal assumption about the floating point resolution that is not met in the end.

* [x] `test_no_chunks`
* [x] `test_reductions_1D[i4]`
* [x] `test_reductions_2D[i4]`
* [x] `test_arg_reductions[dfunc0-func0]`
* [x] `test_arg_reductions[dfunc1-func1]`
* [x] `test_arg_reductions[dfunc2-func2]`
* [x] `test_arg_reductions[dfunc3-func3]`
* [x] `test_nanarg_reductions[dfunc0-func0]`
* [x] `test_nanarg_reductions[dfunc1-func1]`
* [x] `test_reductions_2D_nans`
* [x] `test_reductions_with_negative_axes`
* [x] `test_nan`
* [ ] `test_cov_corr_stable`

**Comment by mrocklin** [comment]
@QuLogic thank you for reporting.  I agree with your assessment that most of these are probably just sloppiness in assuming float64.  I'll take a look sometime tomorrow.

**Comment by QuLogic** [comment]
* For `test_no_chunks`, NumPy is correct and dask is wrong when checked against a `decimal` implementation, and it passes the line before which is almost the same, so a `keepdims=True` bug of some sort?
* For `test_reductions_1D` / `test_reductions_2D` / `test_reductions_2D_nans`, the dask variance is completely off, but maybe this is related to the previous one.
* For `test_arg_reductions` / `test_nanarg_reductions` / `test_reductions_with_negative_axes` / `test_nan`, it appears that dask's `arg*` functions return a different dtype than NumPy's; probably not a big deal.

Here's a better log for the last error:
```
_____________________________ test_cov_corr_stable _____________________________
    @pytest.mark.slow
    def test_cov_corr_stable():
        df = pd.DataFrame(np.random.random((20000000, 2)) * 2 - 1, columns=['a', 'b'])
        ddf = dd.from_pandas(df, npartitions=50)
>       assert_eq(ddf.cov(split_every=8), df.cov())
dask/dataframe/tests/test_dataframe.py:1945: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
dask/dataframe/utils.py:509: in assert_eq
    tm.assert_frame_equal(a, b, **kwargs)
/usr/lib/python2.7/site-packages/pandas/util/testing.py:1313: in assert_frame_equal
    obj='DataFrame.iloc[:, {0}]'.format(i))
/usr/lib/python2.7/site-packages/pandas/util/testing.py:1181: in assert_series_equal
    obj='{0}'.format(obj))
pandas/src/testing.pyx:59: in pandas._testing.assert_almost_equal (pandas/src/testing.c:3783)
    ???
pandas/src/testing.pyx:173: in pandas._testing.assert_almost_equal (pandas/src/testing.c:2987)
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
obj = 'DataFrame.iloc[:, 0]'
message = 'DataFrame.iloc[:, 0] values are different (50.0 %)'
left = '[0.333301151028, 1.34151399883e-05]'
right = '[0.333301804485, 1.3488365727e-05]', diff = None
    def raise_assert_detail(obj, message, left, right, diff=None):
        if isinstance(left, np.ndarray):
            left = pprint_thing(left)
        if isinstance(right, np.ndarray):
            right = pprint_thing(right)
    
        msg = """{0} are different
    
    {1}
    [left]:  {2}
    [right]: {3}""".format(obj, message, left, right)
    
        if diff is not None:
            msg = msg + "\n[diff]: {diff}".format(diff=diff)
    
>       raise AssertionError(msg)
E       AssertionError: DataFrame.iloc[:, 0] are different
E       
E       DataFrame.iloc[:, 0] values are different (50.0 %)
E       [left]:  [0.333301151028, 1.34151399883e-05]
E       [right]: [0.333301804485, 1.3488365727e-05]
/usr/lib/python2.7/site-packages/pandas/util/testing.py:1018: AssertionError
```

**Comment by mrocklin** [comment]
array and dataframe test suites pass for me under a 32 bit miniconda install with https://github.com/dask/dask/pull/2023

**Comment by QuLogic** [comment]
#2023 fixes all but the last test failure for me.

**Comment by mrocklin** [comment]
The last failure looks like it's probably actually just a 32bit precision error.  Not sure though.  @jcrist do have a decent understanding of how precise the cov/corr algorithm is supposed to be?  It looks like we're accurate down to the 6th decimal place on 32 bit systems.  Any concern here?

**Comment by QuLogic** [comment]
BTW, this test uses `np.random` without a seed; how do you ensure that it's not just picking up an unfortunate set of values? I guess it is a pretty large set of numbers at least...

**Comment by QuLogic** [comment]
Just a ping; it seems that `test_cov_corr_stable` is still off with 0.18.1.

**Comment by QuLogic** [comment]
`test_cov_corr_stable` remains [incorrect with 0.19.4](https://koji.fedoraproject.org/koji/taskinfo?taskID=30361488):
```
>       raise AssertionError(msg)
E       AssertionError: DataFrame.iloc[:, 0] are different
E       
E       DataFrame.iloc[:, 0] values are different (50.0 %)
E       [left]:  [0.9999970104972555, 1.707463499923931e-05]
E       [right]: [1.0, 1.7058029018471343e-05]
```

**Comment by mrocklin** [comment]
>  BTW, this test uses np.random without a seed; how do you ensure that it's not just picking up an unfortunate set of values? I guess it is a pretty large set of numbers at least...

Yeah, I just ran this on my 64 bit system 100 times with 10x fewer numbers and it was fine.  I do agree though that we should use a seed and reduce the size of this test though.  It's over 10s long.  

> We should probably be working with 64 bit values internally throughout anyway. I suspect there's probably a place or two where this isn't happening.

Agreed with this assessment.  If anyone wants to track this down my recommendation would be to create a smaller example and then look at the intermediate values within the `cov_corr_chunk`, `cov_corr_combine`, and `cov_corr_agg` functions during a run and see if any of them are allocating 32-bit values.  It *looks* like we've been careful within those functions to specify dtypes, but presumably we've missed something.

**Comment by QuLogic** [comment]
The other issue is #2507, but the 32-bit tests that are left there are also failing on other arches.

**Comment by QuLogic** [comment]
So I don't now when exactly, but it looks like the final failure in `test_cov_core_stable` is fixed in 1.1.3: https://koji.fedoraproject.org/koji/taskinfo?taskID=33258178 (ignore the final failure and just look at the pytest output in the build log.)
